scripts/lib/dependencies.ts

```diff
@@ -128,45 +100,207 @@ export async function writeDependencyVersions(path: string, depVersions: Depende
 
-export function IsGitHubDependency(dependency: Dependency): dependency is GitHubDependency {
-  if (!('githubOwner' in dependency) || typeof dependency.githubOwner !== 'string') {
-    return false;
+  /** The current version. */
+  abstract currentVersion: Promise<Version>;
+
+  /** The newest version that can be upgraded to. */
+  get latestVersion(): Promise<Version> {
+    return (async() => {
+      const availableVersions = await this.getAvailableVersions();
+
+      return availableVersions.reduce((version1, version2) => {
+        return this.rcompareVersions(version1, version2) < 0 ? version1 : version2;
+      });
+    })();
   }
-  if (!('githubRepo' in dependency) || typeof dependency.githubRepo !== 'string') {
-    return false;
+
+  /** Whether we can upgrade. */
+  get canUpgrade(): Promise<boolean> {
+    return (async() => {
+      const current = await this.currentVersion;
+      const latest = await this.latestVersion;
+      const compare = this.rcompareVersions(current, latest);
+
+      if (compare < 0) {
+        throw new Error(`${ this.name } at ${ current }, higher than latest version ${ latest }`);
+      }
+
+      return compare > 0;
+    })();
   }
-  if (!('versionToTagName' in dependency) || typeof dependency.versionToTagName !== 'function') {
-    return false;
+
+  /**
+   * Update the version manifest (e.g. `dependencies.yaml`) for this dependency,
+   * in preparation for making a pull request.
+   * @returns The set of files that have been modified.
+   */
+  abstract updateManifest(newVersion: Version): Promise<Set<string>>;
+
+  /**
+   * Compare the two versions.  The return value is:
+   * Value | Description
+   * --- | ---
+   * -1 | `version1` is higher
+   * 0 | `version1` and `version2` are equal
+   * 1 | `version2` is higher
+   *
+   * The default implementation compares version strings that look like `0.1.2.rd3????`.
+   * Note that anything after the number after `rd` is ignored.
+   */
+  rcompareVersions(version1: Version, version2: Version): -1 | 0 | 1 {
+    if (typeof version1 !== 'string' || typeof version2 !== 'string') {
+      throw new TypeError(`default rcompareVersions only handles string versions (got ${ version1 } / ${ version2 })`);
+    }
+
+    const semver1 = semver.coerce(version1);
+    const semver2 = semver.coerce(version2);
+
+    if (semver1 === null || semver2 === null) {
+      throw new Error(`One of ${ version1 } and ${ version2 } failed to be coerced to semver`);
+    }
+
+    if (semver1.raw !== semver2.raw) {
+      return semver.rcompare(semver1, semver2);
+    }
+
+    // If the two versions are equal, assume we have different build suffixes
+    // e.g. "0.19.0.rd5" vs "0.19.0.rd6"
+    const [, match1] = /^\d+\.\d+\.\d+\.rd(\d+)$/.exec(version1) ?? [];
+    const [, match2] = /^\d+\.\d+\.\d+\.rd(\d+)$/.exec(version2) ?? [];
+
+    if (!match1 && !match2) {
+      // Neither have .rd suffix; treat as equal.
+      return 0;
```

Comment:
Does this mean this fallback assumes equality, even though versions like "1.0.0" vs "1.0.1" are not equal?